app/src/main/java/tech/relaycorp/letro/repository/GatewayRepository.kt

```diff
@@ -32,10 +44,45 @@ class GatewayRepository @Inject constructor(@ApplicationContext var context: Con
             Awala.setUp(context)
             try {
                 GatewayClient.bind()
-                _gatewayAvailabilityDataModel.emit(GatewayAvailabilityDataModel.Available)
             } catch (exp: GatewayBindingException) {
                 _gatewayAvailabilityDataModel.emit(GatewayAvailabilityDataModel.Unavailable)
+                return@launch
+            } finally {
+                registerFirstPartyEndpointIfNeeded()
+                importThirdPartyEndpointIfNeeded()
+                _gatewayAvailabilityDataModel.emit(GatewayAvailabilityDataModel.Available)
             }
         }
     }
+
+    private suspend fun registerFirstPartyEndpointIfNeeded() {
+        if (_firstPartyEndpointNodeId != null) return
+
+        val endpoint = FirstPartyEndpoint.register()
+        preferencesDataStoreRepository.setFirstPartyEndpointAddress(endpoint.nodeId)
+    }
+
+    private suspend fun importThirdPartyEndpointIfNeeded() {
```

Comment:
There's one more thing I forgot to mention on the Jira issue: We should, at some point before sending the first message to the server, authorise the server to send _us_ messages.

I think the quickest/simplest way to do that is to do it preemtively when importing the third party endpoint. Here's basically what we need to do:

```kotlin
// Create the Parcel Delivery Authorisation (PDA)
val auth = serverFirstPartyEndpoint.authorizeIndefinitely(serverThirdPartyEndpoint)
// Send it to the server
val authMessage = OutgoingMessage.build(
  "application/vnd+relaycorp.awala.pda-path",
  auth,
  serverFirstPartyEndpoint,
  serverThirdPartyEndpoint,
  ZonedDateTime.now().plusDays(90), // Expiry date
)
GatewayClient.sendMessage(authMessage)
```

Reply:
Done